src/lua/decimal.c

```diff
@@ -81,14 +82,21 @@ ldecimal_##name(struct lua_State *L) {						\
 
 uint32_t CTID_DECIMAL;
 
-/** Push a new decimal on the stack and return a pointer to it. */
 decimal_t *
-luaT_pushdecimal(struct lua_State *L)
+luaT_newdecimal(struct lua_State *L)
 {
 	decimal_t *res = luaL_pushcdata(L, CTID_DECIMAL);
 	return res;
 }
 
+decimal_t *
+luaT_pushdecimal(struct lua_State *L, const decimal_t *dec)
+{
+	decimal_t *res = luaT_newdecimal(L);
+	memcpy(res, dec, sizeof(decimal_t));
```

Comment:
> * I use `memcpy()` instead of just `*a = *b` in `luaT_pushdecimal()` to
  copy the padding byte content. Who knows, maybe this not-so-legal way
  to hold extra information may be crucial for some use case or will
  allow us to add one field into the structure.

Could you explain this, please? How does `memcpy` help to copy the padding here? As far as I understand, this is identical to `*res = *dec`. And the assignment will copy everything. If `decimal_t` is defined with padding, both assignment and `memcpy` will copy it, and if it's defined without padding, neither of these ops will copy it.

Reply:
> ### 6.2.6 Representations of types
> #### 6.2.6.1 General
>
> <...>
>
> 6. When a value is stored in an object of structure or union type, including in a member
object, the bytes of the object representation that correspond to any padding bytes take
unspecified values.<sup>42)</sup> <...>
>
> <...>
>
> 42\) Thus, for example, structure assignment need not copy any padding bits.

https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf

----

We have one byte padding between the `uint8_t` field and the `uint16_t[]` array. It is due to natural alignment: `uint16_t[]` array starts from an even address, 16 bits boundary. The `uint8_t` field itself always start from an even address too. It is easier to illustrate this way:

```c
typedef struct {      /* Must be aligned to 4 bytes boundary. */
	int32_t digits;   /* Must be aligned to 4 bytes boundary. */
	int32_t exponent; /* Must be aligned to 4 bytes boundary. */
	uint8_t bits;     /* Aligned to 4 bytes boundary, because
                       * follows the previous aligned field. */
	/* One byte padding, becase `lsu` must be aligned. */
	uint16_t lsu[13]; /* Must be aligned to 2 bytes boundary. */
} decNumber;
```

----

My explanation of using `memcpy()` can't be an argument to prefer `memcpy()` over the structure assignment. It is just explanation why I choose it from two almost identical ways to copy the object. If you have *any* argument to prefer `*res = *src` over `memcpy()`, I can't beat it and will change to what you'll suggest. 'To allow the hack' is not an argument, just my tiny wish.

Reply:
Thanks! I see now. I thought you were talking about our padding `uint8_t pad[64]` or something similar. Hence my confusion.
Anyway, nice explanation about the natural padding.
I'm fine with `malloc` here.